src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Areas/Identity/RevalidatingAuthenticationStateProvider.cs

```diff
@@ -1,98 +1,59 @@
+        private readonly IdentityOptions _options;
 
-        public RevalidatingAuthenticationStateProvider(
+        public RevalidatingIdentityAuthenticationStateProvider(
+            ILoggerFactory loggerFactory,
             IServiceScopeFactory scopeFactory,
-            SignInManager<TUser> circuitScopeSignInManager,
-            ILogger<RevalidatingAuthenticationStateProvider<TUser>> logger)
+            IOptions<IdentityOptions> optionsAccessor)
+            : base(loggerFactory, RevalidationInterval)
         {
-            var initialUser = circuitScopeSignInManager.Context.User;
-            _currentAuthenticationStateTask = Task.FromResult(new AuthenticationState(initialUser));
             _scopeFactory = scopeFactory;
-            _logger = logger;
-
-            if (initialUser.Identity.IsAuthenticated)
-            {
-                _ = RevalidationLoop();
-            }
+            _options = optionsAccessor?.Value ?? new IdentityOptions();
         }
 
-        public override Task<AuthenticationState> GetAuthenticationStateAsync()
-            => _currentAuthenticationStateTask;
-
-        private async Task RevalidationLoop()
+        protected override async Task<bool> ValidateAuthenticationStateAsync(AuthenticationState authenticationState)
         {
-            var cancellationToken = _loopCancellationTokenSource.Token;
-
-            while (!cancellationToken.IsCancellationRequested)
+            // Get the user manager from a new scope to ensure it fetches fresh data
+            using (var scope = _scopeFactory.CreateScope())
             {
-                try
-                {
-                    await Task.Delay(RevalidationInterval, cancellationToken);
-                }
-                catch (TaskCanceledException)
-                {
-                    break;
-                }
-
-                var isValid = await CheckIfAuthenticationStateIsValidAsync();
-                if (!isValid)
-                {
-                    // Force sign-out. Also stop the revalidation loop, because the user can
-                    // only sign back in by starting a new connection.
-                    var anonymousUser = new ClaimsPrincipal(new ClaimsIdentity());
-                    _currentAuthenticationStateTask = Task.FromResult(new AuthenticationState(anonymousUser));
-                    NotifyAuthenticationStateChanged(_currentAuthenticationStateTask);
-                    _loopCancellationTokenSource.Cancel();
-                }
+                var userManager = scope.ServiceProvider.GetRequiredService<UserManager<TUser>>();
+                return await ValidateSecurityStampAsync(userManager, authenticationState.User);
             }
         }
 
-        private async Task<bool> CheckIfAuthenticationStateIsValidAsync()
+        private async Task<bool> ValidateSecurityStampAsync(UserManager<TUser> userManager, ClaimsPrincipal principal)
```

Comment:
This logic is copied directly from [the equivalent in `SignInManager`](https://github.com/aspnet/AspNetCore/blob/master/src/Identity/Core/src/SignInManager.cs#L260), but made to look simpler and more approachable. Functionally, it should be identical.